File: misc/level1/04-maxchar/main.go
package main

import (
	"fmt"
	"strings"
)

func maxchar(str string) string {
	strMap := make(map[string]int)

	sArray := strings.Split(str, "")

	max := 0
	maxC := ""

	for i := 0; i < len(sArray); i++ {
		_, ok := strMap[sArray[i]]

		if ok {
			strMap[sArray[i]]++
		} else {
			strMap[sArray[i]] = 1
		}

		if strMap[sArray[i]] > max {
			max = strMap[sArray[i]]
			maxC = sArray[i]
		}
	}

	return maxC
}

func main() {
	fmt.Println(maxchar("abccccccd"))
	fmt.Println(maxchar("apple 123111"))
}

File: misc/level1/04-maxchar/main.rb
# def maxchar(s)
#   hash = {}
  
#   s.split('').each do |c|
#     unless hash.has_key? c
#       hash[c] = 1
#     else
#       hash[c] += 1
#     end
#   end
#   p hash.max_by{ |k,v| v }.first
# end


def maxchar(s)
  hash = {}
  max = 0
  maxc = nil
  s.split('').each do |c|

    hash[c] = (hash[c] == nil) ?  1 : (hash[c]+=1) # ternary expression

    if hash[c] > max
      max = hash[c]
      maxc = c 
    end
  end

  maxc
end

p maxchar("abccccccd") #== "c"
p maxchar("apple 123111") #== "1"
